### ISSUE TITLE
Extend `DoNotIgnoreMethodResultsAnalyzer` (CA1806) for Task-returning non-async methods

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn-analyzers#6181]
```csharp
using System.Threading.Tasks;

public class C
{
    public void M()
    {
        Delay(); // suggestion is to have a warning here.

        var x = Delay(); // no warning.
        DelayAsync(); // no warning.
    }
    
    public Task Delay() => Task.Delay(1000);
    public async Task DelayAsync() => Task.Delay(1000);
}
```

**Comment by Youssef1313** [comment]
@mavasani I'm not sure if this should go to dotnet/runtime for review or not.
cc @stephentoub

**Comment by mavasani** [comment]
@Youssef1313 I believe https://github.com/dotnet/runtime/issues/34098 has the newer design, which involves a new attribute to mark methods whose value should not be ignored.

**Comment by Youssef1313** [comment]
Thanks @mavasani, though for my case, it was accidental that a method return was void and changed to Task, and callsites didn't await. So, I mean such accidental cases will not have the attribute until the bug is discovered later. 

The request here is to be able to detect all usages of this pattern.

**Comment by stephentoub** [comment]
This would have a _huge_ number of false positives.  Task.Run, for example, is frequently used to launch work in a fire-and-forget manner.

**Comment by Youssef1313** [comment]
@stephentoub Should I close then? or will special-casing `Task.Run` solve the issue?

**Comment by stephentoub** [comment]
It's not just Task.Run. We should close this. Thanks.

**Comment by Youssef1313** [comment]
@stephentoub Thinking again about this, would it make sense to discard the result to communicate the intention of fire-and-forget?

For example, the following produces a *compiler warning*:

```csharp
using System.Threading.Tasks;

public class C
{
    public async void M()
    {
        await Task.Delay(1000);

        // Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call.
        Delay();
    }
    
    public Task Delay() => Task.Delay(1000);
}
```

That looks somewhat similar to what I'm suggesting. The compiler warning above might be false positive if the intention is fire-and-forget? Assigning the returned `Task` to a discard silences the compiler warning above. Would that make sense for the analyzer I'm suggesting here?

**Comment by stephentoub** [comment]
> @stephentoub Thinking again about this, would it make sense to discard the result to communicate the intention of fire-and-forget?

That's a style concern.  I'd like to avoid our analyzers getting into that business.

**Comment by jlaanstra** [comment]
Would it be useful to allow a configuration option for this as part of additional_use_results_methods? I don't think the current syntax allows me to specify "all Task/Task<T>/ValueTask/ValueTask<T> returning methods". I understand there are more advanced uses with custom types providing GetAwaiter but the majority of methods use the above types and requiring explicit discard for those is super helpful to make sure all exceptions are observed.
